### ISSUE TITLE
API command POST /container/create ignoring VolumeDriver config from Docker-py

### ISSUE THREAD

**Issue by mxylin** [docker/swarm#2669]
### Summary

**Symptom:** When creating containers with named volumes using Docker SDK for Python (https://github.com/docker/docker-py/), the named volumes will always be created using *local* driver whether a volume driver is specified or not. 
 
**Cause:** VolumeDriver config generated by current version (2.2.1) of Docker SDK for Python  to create/run a container is ignored/filtered by Swarm manager.

#### Software versions tested

* **Swarm:** 1.2.6, 1.2.5, 1.2.4
* **Docker Engine:** 17.03.0-ce
* **Docker-py:** 2.2.1, 2.1.0, 1.10.6, 1.8.1

#### Swarm manager `docker info`

```
Containers: 4
 Running: 0
 Paused: 0
 Stopped: 4
Images: 8
Server Version: swarm/1.2.6
Role: primary
Strategy: spread
Filters: health, port, containerslots, dependency, affinity, constraint, whitelist
Nodes: 2
 node1: 172.17.177.21:2376
  └ ID: <REDACTED>
  └ Status: Healthy
  └ Containers: 2 (0 Running, 0 Paused, 2 Stopped)
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.017 GiB
  └ Labels: kernelversion=4.4.0-66-generic, operatingsystem=Ubuntu 16.04.1 LTS, storagedriver=aufs
  └ UpdatedAt: 2017-03-30T07:51:22Z
  └ ServerVersion: 17.03.0-ce
 node2: 172.17.177.22:2376
  └ ID: <REDACTED>
  └ Status: Healthy
  └ Containers: 2 (0 Running, 0 Paused, 2 Stopped)
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.017 GiB
  └ Labels: kernelversion=4.4.0-66-generic, operatingsystem=Ubuntu 16.04.1 LTS, storagedriver=aufs
  └ UpdatedAt: 2017-03-30T07:50:56Z
  └ ServerVersion: 17.03.0-ce
Plugins:
 Volume:
 Network:
Swarm:
 NodeID:
 Is Manager: false
 Node Address:
Kernel Version: 4.4.0-66-generic
Operating System: linux
Architecture: amd64
CPUs: 2
Total Memory: 2.035 GiB
Name: <REDACTED>
Docker Root Dir:
Debug Mode (client): false
Debug Mode (server): false
Experimental: false
Live Restore Enabled: false
```

### Steps to reproduce this issue

1. Set up a Swarm cluster with volume plugins configured for each node. For example, install convoy (https://github.com/rancher/convoy) on all Swarm nodes.

2. Install Docker SDK for Python on the Swarm manager.
   * use `pip install docker` to install latest (2.x) version
   * use `pip install docker-py` to install 1.x version

3. Set up environment variable DOCKER_HOST to point to the Swarm manager.
   * `export DOCKER_HOST=localhost:2375`

4. Run the following Python code:
```python
# If testing 2.x version of Docker SDK for Python
import docker
client = docker.from_env()
container = client.containers.create(image='ubuntu', command='uname', name='ubuntu-test', volume_driver='convoy', volumes={'testvol': {'bind': '/mnt/vol', 'mode': 'rw'}})
```
```python
# If testing 1.x version of Docker SDK for Python
import docker
client = docker.from_env()
container = client.create_container(image='ubuntu', command='uname', name='ubuntu-test', volume_driver='convoy', volumes=['/mnt/vol'],host_config=client.create_host_config(binds=['testvol:/mnt/vol']))
```

5. Check for created container and volume on Swarm nodes using `docker ps -a` and `docker volume ls`.

#### Result:

1. A new container is created and scheduled to one of the Swarm nodes, as expected.
```
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
6cd32cd9c739        ubuntu              "uname"             6 seconds ago       Created                                 node2/ubuntu-test
```

2. A new named volume is created using *local* driver instead of specified *convoy* driver. This is an **unexpected** result.
```
$ docker volume ls
DRIVER              VOLUME NAME
local               node2/testvol
```

### Observations

1. The `VolumeDriver` was moved from `config` to `HostConfig` since Docker Engine v1.9 (API version: v1.21).
   * [Engine API v1.21 changes](https://docs.docker.com/engine/api/version-history/#v121-api-changes)
   * Related PR: docker/docker#15798

2. Current version of Docker SDK for Python places VolumeDriver in `config` part.

3. Current version of Swarm ignores VolumeDriver in `config` part.

#### HTTP request dump of `client.containers.create()` from Docker SDK for Python v2.2.1:
```
POST /v1.26/containers/create?name=ubuntu-test HTTP/1.1
Host: localhost:2375
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 314
Content-Type: application/json
User-Agent: docker-sdk-python/2.2.1

{
	"Tty": false,
	"NetworkDisabled": false,
	"Image": "ubuntu",
	"Cmd": [
		"uname"
	],
	"StdinOnce": false,
	"HostConfig": {
		"Binds": [
			"testvol:/mnt/vol:rw"
		],
		"NetworkMode": "default"
	},
	"AttachStdin": false,
	"Volumes": {
		"/mnt/vol": {}
	},
	"VolumeDriver": "convoy",
	"AttachStderr": true,
	"AttachStdout": true,
	"OpenStdin": false
}
```

#### HTTP request dump of `client.create_container()` from Docker SDK for Python v1.10.6:
```
POST /v1.24/containers/create?name=ubuntu-test HTTP/1.1
Host: localhost:2375
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 311
Content-Type: application/json
User-Agent: docker-py/1.10.6

{
	"Tty": false,
	"NetworkDisabled": false,
	"Image": "ubuntu",
	"Cmd": [
		"uname"
	],
	"StdinOnce": false,
	"HostConfig": {
		"Binds": [
			"testvol:/mnt/vol"
		],
		"NetworkMode": "default"
	},
	"AttachStdin": false,
	"Volumes": {
		"/mnt/vol": {}
	},
	"VolumeDriver": "convoy",
	"AttachStderr": true,
	"AttachStdout": true,
	"OpenStdin": false
}
```

### Proposed fix

I've made a commit mxylin@cfb6cb2916c7d2ce1b1a3638b0bf9cc65f4f7d57 which fixes this issue by adding backward compatibility for VolumeDriver in ContainerConfig.

**Comment by mxylin** [comment]
The Docker image for mxylin@cfb6cb2916c7d2ce1b1a3638b0bf9cc65f4f7d57 is now available at https://hub.docker.com/r/xylin/swarm/

**Comment by CodeJuan** [comment]
y

**Comment by dongluochen** [comment]
@mxylin Sorry for our delay. https://github.com/docker/docker-py/pull/1547 should have fixed your issue, right?

**Comment by mxylin** [comment]
Hi @dongluochen thanks for the reminder.

I've just ran some tests and the issue is fixed for `docker.APIClient` but not `docker.DockerClient`.
For details please refer to https://github.com/docker/docker-py/pull/1547#issuecomment-294398121.

Thanks!

**Comment by mxylin** [comment]
Fixed in docker/docker-py/pull/2288

Related issue: docker/docker-py/issues/2271
